tests/test_mailmerge.py

```diff
@@ -0,0 +1,262 @@
+"""Mailmerge unit tests."""
+import os
+import email
+import cd
+import markdown
+import mailmerge
+
+
+# NOTE: Python 2.x mock lives in a different place
+try:
+    from unittest import mock
+except ImportError:
+    import mock
+
+
+# We're going to use mock_SMTP because it mimics the real SMTP library
+# pylint: disable=invalid-name
+
+
+# Directories containing test input files
+TESTDIR = os.path.dirname(__file__)
+TESTDATA = os.path.join(TESTDIR, "testdata")
+
+
+def test_stdout(capsys):
+    """Verify stdout and stderr.
+
+    pytest docs on capturing stdout and stderr
+    https://pytest.readthedocs.io/en/2.7.3/capture.html
+    """
+    mailmerge.api.main(
+        database_filename=os.path.join(TESTDATA, "simple_database.csv"),
+        template_filename=os.path.join(TESTDATA, "simple_template.txt"),
+        config_filename=os.path.join(TESTDATA, "server_open.conf"),
+        no_limit=True,
+    )
+
+    # Verify mailmerge output
+    stdout, stderr = capsys.readouterr()
+    assert stderr == ""
+    assert ">>> message 0" in stdout
+    assert ">>> sent message 0 DRY RUN" in stdout
+    assert ">>> message 1" in stdout
+    assert ">>> sent message 1 DRY RUN" in stdout
+
+
+@mock.patch('smtplib.SMTP')
+def test_smtp(mock_SMTP):
+    """Verify SMTP library calls."""
+    mailmerge.api.main(
+        database_filename=os.path.join(TESTDATA, "simple_database.csv"),
+        template_filename=os.path.join(TESTDATA, "simple_template.txt"),
+        config_filename=os.path.join(TESTDATA, "server_open.conf"),
+        limit=1,
+        dry_run=False,
+    )
+
+    # Mock smtp object with function calls recorded
+    smtp = mock_SMTP.return_value
+    assert smtp.sendmail.call_count == 1
+
+
+@mock.patch('smtplib.SMTP')
+def test_cc_bcc(mock_SMTP):
```

Comment:
@seshrs I thought you might be interested to see how the `unittest.mock` library works. You originally wrote this testcase, and I recently refactored it to mock the SMTP library and avoid network requests.

Reply:
Cool!